File: backend/app/schemas/me.py
"""
Bootstrap context schemas.

GET /me/context is the single call that tells the client who the actor is,
which tenants they belong to, and where to send them.

It exists to make three states distinguishable that the pre-ARCH-01 frontend
could not tell apart. OnboardingGuard called getWorkspace() and treated any
falsy result as "no workspace", so an expired token and a genuinely
membership-less user produced the same signal — and session expiry sent people
to "Create My Workspace" instead of the login page. Removal from a workspace
did the same, which is how removed members ended up founding phantom
organizations.

Now:
    401                      -> token invalid or expired    -> /login
    200, organizations: []   -> authenticated, no tenant    -> /onboarding
    200, organizations: [..] -> normal                      -> the workspace
"""

from __future__ import annotations

from uuid import UUID

from pydantic import BaseModel, ConfigDict, EmailStr, Field

from app.models.organization import OrganizationRole, OrganizationStatus
from app.schemas.workspace import WorkspaceSummary


class MeUser(BaseModel):
    """
    The authenticated actor.

    ARCH-05 Step 5: carries `display_name` and `timezone` now that `users`
    has them — this is the payload the sidebar and header render from on
    every session start, and "the sidebar renders `user?.email`" was one of
    A.2.4's own named symptoms. `locale` is intentionally absent: nothing at
    boot time needs it yet, and `GET /me/profile` (`UserProfileResponse`,
    `app/schemas/user.py`) is the fuller, deliberately-opened profile
    contract for a settings screen. Keeping this projection minimal is the
    same reasoning `UserProfileResponse`'s docstring gives for NOT reusing
    this class there: a field added here for bootstrap reasons should not
    show up on the settings page for no reason, and vice versa.
    """
    id: UUID
    email: EmailStr
    is_active: bool
    display_name: str | None = None
    timezone: str | None = None

    model_config = ConfigDict(from_attributes=True)


class OrganizationMembershipSummary(BaseModel):
    """
    One organization the actor belongs to, with their standing and reachable
    workspaces.

    Workspaces are embedded rather than fetched per organization so the client
    can render the full switcher from a single response.
    """
    organization_id: UUID
    organization_slug: str
    organization_name: str
    organization_status: OrganizationStatus
    role: OrganizationRole
    workspaces: list[WorkspaceSummary] = Field(default_factory=list)


class MeContextResponse(BaseModel):
    """
    Everything the client needs on boot, in one round trip.

    default_organization_id and default_workspace_id resolve the landing
    destination server-side. Returning null for both alongside an empty
    organizations list is the unambiguous "send this user to onboarding"
    signal — reachable only with a valid session, which is precisely the
    distinction that was missing before.
    """
    user: MeUser
    organizations: list[OrganizationMembershipSummary] = Field(
        default_factory=list
    )
    default_organization_id: UUID | None = None
    default_workspace_id: UUID | None = None
    requires_onboarding: bool = Field(
        ...,
        description=(
            "True when the actor belongs to no organization and should be "
            "sent to tenant creation. Never inferred from a failed request."
        ),
    )

File: frontend/src/services/api/me.ts
/**
 * Actor-scoped API service for FlowPilot AI.
 *
 * Answers "who am I and what can I reach" without naming a tenant.
 *
 * getMeContext is the application bootstrap and the fix for a specific defect.
 * The pre-ARCH-01 OnboardingGuard called getWorkspace() and treated any falsy
 * result as "no workspace", so an expired token and a genuinely
 * membership-less user produced the same signal — and session expiry sent
 * people to "Create My Workspace" instead of the login page. Removal from a
 * workspace did the same, which is how removed members ended up founding
 * phantom organizations.
 *
 * The states are now distinguishable at the transport layer:
 *
 *   throws ApiError(401, UNAUTHORIZED)  -> session gone       -> /login
 *   resolves, requires_onboarding true  -> no tenant          -> onboarding
 *   resolves, organizations populated   -> normal             -> workspace
 *
 * A rejected promise and a resolved one can no longer be confused, which is
 * exactly what the old boolean check could not achieve.
 */

import apiClient from "@/services/api/client";
import { ME_ENDPOINTS } from "@/services/api/endpoints";

import type {
  MeContext,
  Organization,
  WorkspaceSummary,
} from "@/types/tenancy";

/**
 * Returns the complete bootstrap context in one round trip.
 *
 * Rejects with ApiError(401) when the session is invalid. Never resolves with
 * an "unauthenticated" shape — that distinction is the point of the endpoint.
 */
export const getMeContext = async (): Promise<MeContext> => {
  const response = await apiClient.get<MeContext>(ME_ENDPOINTS.context, {
    headers: { Accept: "application/json" },
  });
  return response.data;
};

/**
 * Returns every organization the actor actively belongs to.
 *
 * Multiple results are ordinary. The pre-ARCH-01 backend could not represent
 * this at all: a second membership raised MultipleResultsFound and returned
 * HTTP 500 on every subsequent request.
 */
export const getMyOrganizations = async (): Promise<Organization[]> => {
  const response = await apiClient.get<Organization[]>(
    ME_ENDPOINTS.organizations,
    { headers: { Accept: "application/json" } },
  );
  return response.data;
};

/**
 * Returns every workspace the actor holds an EXPLICIT grant on.
 *
 * Deliberately excludes workspaces reachable only through organization-level
 * derived elevation. An organization admin of a large tenant would otherwise
 * receive every workspace in it — correct, but not what a personal workspace
 * list means. Use getMeContext for the complete, grouped view.
 */
export const getMyWorkspaces = async (): Promise<WorkspaceSummary[]> => {
  const response = await apiClient.get<WorkspaceSummary[]>(
    ME_ENDPOINTS.workspaces,
    { headers: { Accept: "application/json" } },
  );
  return response.data;
};

export const meApi = {
  getMeContext,
  getMyOrganizations,
  getMyWorkspaces,
};

export default meApi;
